Add Sourcery code generation for tests

broomburgo opened the pull request:
This is a first step, in which I used Sourcery to generate code for tests regarding algebraic structures' laws.

I added the Stencil templates to the git index, but not the Sourcery binary: there's a Xcode build phase that runs Sourcery if present, but it shouldn't be necessary because the generated test files are committed. Also, I'm not sure is a good idea to commit the Sourcery binary: anyway, for the script to work the Sourcery executable (along with its app) must be present in a `Sourcery` folder.

I wasn't able to properly use Sourcery to generate `LinuxMain.swift`: I'm still generating it with a template, but apparently Sourcery ignores the "generated" files in a second pass, so I manually added the references to the `allTests` static variables.

closes #17

bkase commented on Templates/Other/LinuxMain.stencil:
```diff
@@ -0,0 +1,14 @@
+
+XCTMain([
+{% for type in types.based.XCTestCase %}
+	testCase({{ type.name }}.allTests),
```
should we use 4 spaces instead of tabs?

broomburgo replied:
Do you mean in the Stencil templates? Why? I don't have strong feelings about this but I indent with tabs in all of the code I write, and the code in Abstract uses tabs. Is there a particular reason to use spaces instead of tabs in this context?

bkase replied:
Oh you're right. Didn't realize -- best to be consistent!

bkase commented on Tests/AbstractTests/BoundedSemilatticeTests.generated.swift:
```diff
@@ -0,0 +1,48 @@
+// Generated using Sourcery 0.7.2 — https://github.com/krzysztofzablocki/Sourcery
```
It would be cool to not commit the generated code with Sourcery -- we could just make it a requirement to build the project. The problem with committing artifacts is that it makes PRs and reading code really messy (because people have to sift through generated code). What do you think?

broomburgo replied:
If the library is to be targeted also to Linux we cannot make that requirement because Sourcery is still not available for Linux. And in general I prefer to commit all the sources, autogenerated or not, so that the library works out-of-the-box, and in the end you don't really need to check the code in files  that start with, in this case `Generated using Sourcery`, but only the relative templates. But in this particular case code generation is only used for tests, so the library is still usable without Sourcery, which is needed only if you want to run tests locally.

If you think it's better to not commit the generated files is fine by me, but what about the Sourcery binary (and companion app)? Do you think we should add those to the git index?

bkase replied:
I think Sourcery's lack of existence in Linux means that we should check these in. You're right.

bkase commented on sourceryTests.sh:
```diff
@@ -0,0 +1,16 @@
+echo "Checking for Sourcery..."
+
+if [ -f "Sourcery/sourcery" ]; then
+	echo "Sourcery: Sources/Abstract + Templates/Tests -> Tests/AbstractTests"
```
could you use space characters here too?

broomburgo replied:
Same as the first question.

broomburgo commented:
I just updated `sourceryTests.sh` to make Sourcery actually consider `.generated` test files during second pass, required to properly generate `LinuxMain.swift`: the first line, starting with `// Generated using Sourcery` must be removed, but I added a third pass to restore the files after `LinuxMain.swift` generation.
